### PR TITLE
API: ``genfromtxt``: Change default of parameter ``dtype`` to None

### PR DESCRIPTION
**Pull request by timrdmr**
Fix issue #24900. This is my first PR. Please give feedback in case I misunderstood the process.

### CODE REVIEW

File: numpy/lib/tests/test_io.py [hunk 1]
```diff
@@ -1827,11 +1827,11 @@ def test_empty_file(self):
             sup.filter(message="genfromtxt: Empty input file:")
             data = TextIO()
             test = np.genfromtxt(data)
-            assert_equal(test, np.array([]))
```
Comment (timrdmr) [inline numpy/lib/tests/test_io.py:1824]: Here, we would have a mismatch between `np.array([], dtype=[])` and `np.array([], dtype=float64)`

**Comment by timrdmr** [comment]
After more failing tests in `doc/source/user/basics.io.genfromtxt.rst`, I am not sure if the default value of the `dtype` parameter  should remain float.

**Comment by timhoffm** [comment]
Fundamentally, I'd say that auto-detection is better than a forced float. But one needs to be cautious with this change due to backward-compatibility. If a user had a file that only contained ints, they would still have gotten a float array. After this change, they will get an int array. This may break their code.

I'm not a core dev, so have no clear view on change policy, but IMHO this would at least need a deprecation / upcoming change notice.

**Comment by ngoldbaum** [comment]
A search for "np.genfromtxt" on github search returns [60,000 Python code results](https://github.com/search?q=%22np.genfromtxt%22+language%3APython&type=code). This makes it a very commonly used function. Looking over the first few pages, in general people tend to not explicitly specify `dtype` so I agree that it's likely making this change will break user code.

It probably does make sense to deprecate not explicitly setting the output dtype. One way forward would be to change the default from `float` to some other sentinel specifically for this purpose that is not a dtype, and when someone leaves it unset the implementation can check for the sentinel and print a deprecation, along with a note that in the future the default dtype will be inferred from the data, and a suggestion to use "dtype=float" if they want the old behavior or "dtype=None" if they want the future behavior.

This doesn't necessarily need to happen as part of NumPy 2.0, deprecations like this can happen in any major NumPy release.

**Comment by timhoffm** [comment]
I think the default `dtype=float` isn't actually the root problem of #24900.

To clarify the original issue: That complex numbers are silently converted to NaN, is the result of the combined default setting `dtype=float, loose=True`. IMHO the second one is much more offending. The user should have to opt into accepting invalid values (errors should never pass silently, unless explicitly silenced). I thus suggest to migrate to `loose=False` via a deprecation to solve #24900.

---

Side note to changing to `dtype=None` by default:

In a quick test, `dtype=None` is slower than `dtype=float` by a factor of 1.6 for a float input. Are we willing to accept this slowdown? The decision between `dtype=None` and `dtype=float` default is purely between speed and convenience (autodetect a suitable type). Both are justifiable options, if starting from scratch.

Given the broad usage `genfromtxt()`, one needs to be very careful with behavior changes. I mildly advise against changing to `dtype=None` because a user-friendly migration would be quite a hassle.

**Comment by ngoldbaum** [comment]
That's a good point about this slowing down loading data that is represented well as floats. And I'm sure for many of the existing usages it does make sense at some level to represent the data as floats, so for those usages the proposed deprecation is unnecessary noise.

**Comment by ngoldbaum** [comment]
> I thus suggest to migrate to loose=False via a deprecation to solve #24900.

I agree - this seems much more likely to catch buggy downstream usages than forcing everyone to specify the output dtype. The deprecation warning (and eventual error) could even mention setting the dtype to `None` to force dtype detection.

**Comment by rossbar** [comment]
Just to note - IMO these days there is very little reason to use `genfromtxt` at all. `loadtxt` was rewritten in C and is now significantly faster. `genfromtxt` didn't get the same treatment largely due to the dtype discovery mechanism which is the point of contention here. Really that's `genfromtext`'s only remaining real "feature" relative to `loadtxt`, and I'd wager the dtype discovery is nowhere near as good as `pd.read_csv`.

Just dropping this in there for context - I agree that things should be fixed so that existing usage of `genfromtext` is correct. I've also changed the title/tag for this one since it deals with changing API. I suspect it's possible to fix #24900 without changing any API (just fixing the bug in the dtype determination logic) but I haven't looked into it too deeply myself.

**Comment by timhoffm** [comment]
> I suspect it's possible to fix #24900 without changing any API (just fixing the bug in the dtype determination logic) but I haven't looked into it too deeply myself.

Unfortunately not. We already have the capabilities to (1) autodetect types (`dtype=None`) and (2) to raise if trying to convert an invalid string for a type (`loose=False`). But as explained in https://github.com/numpy/numpy/pull/25301#issuecomment-1843864632 the default choice `dtype=float, loose=True` implies "interpret as float and mark all non-float entries as NaN". So the troubling behavior of #24900 is a result of our default choice and sort of by design. To change it, we need to change the default behavior.

**Comment by rossbar** [comment]
> So the troubling behavior of https://github.com/numpy/numpy/issues/24900 is a result of our default choice and sort of by design.

This is a nice summary - I was under the impression that the problem was with dtype discovery (i.e. `dtype=None`). Given that the defaults are `dtype=float` and `loose=True`, I would argue #24900 is a non-issue. The behavior is correct and as-specified.

Of course, discussions about changing default arguments are valid, but I do not think that #24900 is particularly relevant as a motivation for such discussions.

**Comment by rossbar** [comment]
> Thank you for your comments, let me try to summarise:

Thanks @timrdmr , yes I believe your summary captures the situation nicely

> Maybe it is enough to add a note to the documentation.

It probably makes sense to add a docstring example with `loose` since that is `genfromtext`-specific and is otherwise buried in the parameter list (I missed it in my initial evaluation of #24900)

**Comment by seberg** [comment]
Thanks, but it doesn't sound like this is going anywhere and for docs we can open a new PR. (also for this if there is clarity on doing it).

**Pull request closed without merging**